// swc/dist/swc.js
// https://github.com/vasturiano/swc Version 0.2.0. Copyright 2017 Vasco Asturiano.
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.SWC = global.SWC || {})));
}(this, (function (exports) { 'use strict';

class Prop {
	constructor(name, initVal = null, redigest = true, onChange = newVal => {}) {
		this.name = name;
		this.initVal = initVal;
		this.redigest = redigest;
		this.onChange = onChange;
	}
}

function createComponent(config = {}) {
	const props = config.props || [];
	const initFn = config.init || (() => {});
	const updateFn = config.update || (() => {});

	return function() {

		// Holds component state
		let state = {
			initialised: false
		};

		// Component constructor
		function comp(nodeElement) {
			initStatic(nodeElement);
			digest();

			return comp;
		}

		// Getter/setter methods
		props.forEach(prop => {
			comp[prop.name] = getSetProp(prop.name, prop.redigest, prop.onChange);
			state[prop.name] = prop.initVal;
			prop.onChange(prop.initVal);

			function getSetProp(prop, redigest = false,  onChange = newVal => {}) {
				return _ => {
					if (!arguments.length) { return state[prop] }
					state[prop] = _;
					onChange(_);
					if (redigest) { digest(); }
					return comp;
				}
			}
		});

		// Reset all component props to their default value
		comp.resetProps = function() {
			props.forEach(prop => {
				state[prop.name] = prop.initVal;
				prop.onChange(prop.initVal);
			});
			digest();	// Re-digest after resetting props

			return comp;
		};

		//

		function initStatic(nodeElement) {
			initFn(nodeElement, state);
			state.initialised = true;
		}

		function digest() {
			if (!state.initialised) { return; }
			updateFn(state);
		}

		return comp;
	}
}

exports.Prop = Prop;
exports.createComponent = createComponent;

Object.defineProperty(exports, '__esModule', { value: true });

})));
